.github/CHANGELOG.md

```diff
@@ -2,6 +2,66 @@
 
 <h3>New features since last release</h3>
 
+* Adds a decorator `@qml.qfunc_transform` to easily create a transformation
+  that modifies the behaviour of a quantum function.
+  [(#1315)](https://github.com/PennyLaneAI/pennylane/pull/1315)
+
+  For example, consider the following transform, which replaces all
+  `CRX` gates with a sequence of `RX`, `RY`, and `CZ` gates:
+
+  ```python
+  @qml.qfunc_transform
+  def my_transform(tape, x, y):
+      for op in tape.operations + tape.measurements:
+          if op.name == "CRX":
+              wires = op.wires
+              param = op.parameters[0]
+              qml.RX(x * param, wires=wires[1])
+              qml.RY(y * qml.math.sqrt(param), wires=wires[1])
+              qml.CZ(wires=[wires[1], wires[0]])
+          else:
+              op.queue()
+  ```
+
+  We can now apply this transform to any quantum function:
+
+  ```python
+  dev = qml.device("default.qubit", wires=2)
+
+  def ansatz(x):
+      qml.Hadamard(wires=0)
+      qml.CRX(x, wires=[0, 1])
+
+  @qml.qnode(dev)
+  def circuit(param, transform_weights):
+      qml.RX(0.1, wires=0)
+
+      # apply the transform to the ansatz
+      my_transform(*transform_weights)(ansatz)(param)
```

Comment:
This is a very confusing example, or maybe just confusing syntax.   Why are we transforming the first of the two parameters? I would expect something like `my_transform(ansatz)(transform_wieghts, param)`.

Reply:
I agree in that it does come a bit out of left-field 😆 

Funnily enough, my intuition _prior_ to working on this PR was slightly different - I was expecting something along the lines of 

```python
my_transform(ansatz, transform_weights)(param)
```
(in that the `transform_weights` are a dependent quantity when applying the transform).

The reason instead for 

```python
my_transform(transform_weights)(ansatz)(param)
```

Is due almost completely to how decorators work in Python; decorators must always be of the form `callable -> callable`, which causes the above 'restriction'. However, it is also slightly more flexible - you can break it up into the following 'steps':

```python
# 1. create a transform defined by `transform_weights`
specific_transform = my_transform(transform_weights)  
# 2. Apply the transform (callable -> callable) to the qfunc
new_qfunc = specific_transform(new_qfunc)
# 3. evaluate the new, transformed, quantum function
new_qfunc(params)
```

If we decide to drop the decorator functionality, then we could definitely support
```python
my_transform(ansatz, transform_weights)(param)
```

@glassnotes, what do you think?

- Drop the decorator functionality in favour of a clearer 'inline' syntax?
- Keep the decorator functionality at the expense of a slightly more complicated 'inline' syntax?

Reply:
I'm in favour of the decorators because of how much they abstract away in terms of writing the actual transform. However we should definitely add the example you wrote above to the docs, to show the equivalence of the inline version vs. "expanded" version.